### PR TITLE
Fix: decouple V4.1 compressor state from batch order

### LINKED ISSUE #1260: [Feature] Make DeepSeek V4.1 compressor state cache addressing independent of batch rows
## Background

DeepSeek V4.1 inference needs compressor state to survive packed-batch reordering and request admission/removal. The current metadata builder derives `compressor_state_rows` directly from batch-local `request_ids`, conflating request order with persistent storage ownership.

Reviewed source: [metadata.py at cfe35cb](https://github.com/hw-native-sys/pypto-lib/blob/cfe35cbcd3a6fa5b8b1564b02e112b3c9100bb63/models/deepseek_v4_1_flash/metadata.py#L183). This is an inference API improvement identified by source inspection, not a reported device failure. The complete builder is not currently wired into the model execution path; existing kernels already accept independently supplied state addresses.

### Summary

Replace the batch-row-derived compressor-state addressing contract with an engine-managed state cache and state block table. The public compressor interface should accept `query_start_loc + positions + state_block_table + state_cache + token_to_req_indices`. Compute physical state slots inside the kernel; do not require callers to supply redundant `state_rows` or per-token state `slot_mapping`.

### Motivation / Use Case

Ratio-2 compression needs the previous token's FP32 KV projection and gate score when a pair spans two calls. These values are persistent cache state, owned by the request and source layer.

Concrete failure scenario if the current builder is used without external state relocation:

1. Batch rows are `[A, B]`. B has an unfinished pair saved in state row 1.
2. A finishes. The next batch is `[B, C]`; B's next token closes that pair.
3. The builder now assigns B state row 0. B consumes A's old state instead of its own.

Fixed batch-slot affinity or copying every source layer's state during reordering could avoid this, but neither is an explicit current contract. Physical cache addressing should remain independent of batch order.

### Proposed API / Behavior

The state-related inputs, in addition to the existing projection/weight/output arguments, are:

| Input | Shape / meaning |
| --- | --- |
| `query_start_loc` | INT32 `[B + 1]`, packed query boundaries |
| `positions` | Integer `[T]`, absolute token positions |
| `token_to_req_indices` | INT32 `[T]`, token to current batch-row mapping; not a business request ID |
| `state_block_table` | INT32 `[B, 1]`, engine-allocated physical state block per request |
| `state_cache` | FP32 `[num_state_blocks, capacity, 2 * head_dim]`, per-source-layer ring cache; first half KV projection, second half gate score |

For each valid token, the kernel computes:

```python
request_row = token_to_req_indices[token]
block = state_block_table[request_row, 0]
slot = block * capacity + positions[token] % capacity
```

When a chunk starts with a token that closes a pair, read its predecessor from the same block at `(positions[token] - 1) % capacity`. Pairs wholly inside a chunk can read the current projections directly. Ensure historical reads complete before ring writes overwrite their locations, including chunks longer than the ring capacity.

- Keep `token_to_req_indices` as an accepted public input. Define its consistency with `query_start_loc` and use the descriptive name instead of ambiguous `request_ids` at this boundary.
- The engine owns state-block allocation, reassignment, and lifetime. The kernel must never infer physical ownership from a batch row.
- State block tables are separate from compressed-KV/index-cache block tables. Each ratio-2 KV source owns its state cache; document whether physical block numbering is shared across source pools.
- Use the same cache ABI for prefill and decode. Capacity is a declared configuration/shape property, not context length. Ordinary ratio-2 inference can use bounded state; a ring layout leaves room for future verification workloads.
- Define valid-token counts and invalid-block behavior; padded/inactive tokens must not access or mutate state. Idle DP ranks must have a defined zero-work path.
- A fresh request starts with no pending pair. Resuming at an odd position requires the matching predecessor state to be restored or recomputed; KV prefix restoration alone is insufficient.
- Ratio 1 has no recurrent state cache.
- Update the metadata builder, C2A prefill/decode signatures and kernels, Torch references, fixtures, and public documentation together. Preserve numerical compressor semantics.

### Acceptance Criteria

- [ ] Public state interface accepts the five inputs above and has no required `compressor_state_rows` input.
- [ ] Nonidentity state-block allocation and request reordering preserve per-request compressed outputs and state, compared with isolated request execution.
- [ ] Request departure/admission and released-block reuse cannot leak stale pending state into a new request.
- [ ] Odd/even chunk boundaries and prefill-to-decode continuation match an unchunked reference; include chunks longer than ring capacity and ring wraparound.
- [ ] Padded tokens, zero-length requests, and empty local batches perform no invalid state accesses; unrelated blocks remain byte-identical.
- [ ] All ratio-2 source caches follow the same ownership contract; ratio-1 paths need no state inputs.
- [ ] Kernel validation uses nonidentity block mappings, not only block ID equal to batch row.
- [ ] Document that ring storage alone does not implement speculative acceptance/rejection or prefix-state restoration.

### Alternatives Considered

- Keep one pending row with an engine-provided stable slot: sufficient for ordinary sequential inference, but a different storage contract from the proposed ring cache.
- Precompute per-token state `slot_mapping`: valid and used by vLLM, but redundant at our public boundary because the kernel already receives positions, the block table, and token-to-request indices.
- Rename `state_rows` only: insufficient; ownership and storage semantics must change together.

### Additional Context

vLLM's **DeepSeek V4.1-specific** implementation at revision `28ec03bc6e51b17cd9a6b48edc3819e542c2ebb1` registers `CompressorStateCache` with `CircularBufferSpec`. Its metadata builder derives ring `slot_mapping` from the request's state block table and absolute position; the compression kernel receives that mapping as `state_slots`. We adopt the state-cache ownership model while moving slot arithmetic into our compressor kernel.

- [vLLM V4.1 state cache and metadata builder](https://github.com/vllm-project/vllm/blob/28ec03bc6e51b17cd9a6b48edc3819e542c2ebb1/vllm/models/deepseek_v41/compressor.py)
- [vLLM V4.1 state-saving/compression kernel](https://github.com/vllm-project/vllm/blob/28ec03bc6e51b17cd9a6b48edc3819e542c2ebb1/vllm/models/deepseek_v41/common/ops/fused_compress_quant_cache.py)

Related: #1205. Scope is the compressor inference interface and cache correctness; full serving orchestration, Engram, DSpark, and speculative decoding implementation remain separate work.

### PR DESCRIPTION
**Pull request by high-cloud**
- Replace batch-derived state rows with engine-owned block tables and
  per-source FP32 ring caches shared by C2A prefill and decode.
- Preserve historical reads before writing each chunk's final ring rows,
  and guard inactive requests, invalid blocks and empty rank work.
- Update metadata, Torch references, fixtures and cache ABI
  documentation together, including nonidentity state allocations and
  ownership rules.
- Preserve per-rank MX scale layouts in C2A prefill host dispatch.
- Add CPU and two-call kernel regressions for request reordering, block
  reuse, chunk boundaries, ring wraparound and untouched state slots.

Fixes #1260

### CODE REVIEW

File: models/deepseek_v4_1_flash/compressor.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+                    end = pl.read(query_start_loc, [request + 1])
+                    position = pl.read(position_ids, [t])
+                    if t >= start and t < end and position >= 0 and position % 2 == 1:
+                        if t > start:
+                            kv = kv_proj[t - 1 : t, :]
+                            score = score_proj[t - 1 : t, :]
+                        else:
+                            slot = block * STATE_CAPACITY + (position - 1) % STATE_CAPACITY
+                            kv = flat[slot : slot + 1, :HEAD_DIM]
+                            score = flat[slot : slot + 1, HEAD_DIM:]
```
[the remark this answers is not in the document]
Comment (high-cloud) [reply models/deepseek_v4_1_flash/compressor.py]: Fixed in 858c72e: request, block, and ring-slot offsets now explicitly use pl.INDEX in all three compressor stages. Added 20 docstrings for functions touched by this PR, including the regression helpers. Validated with PyPTO main 04f1f510 and its pinned runtime/toolchain: 329 CPU tests, simulator regressions, and A5 compressor, decode, and prefill Full/Reuse cases pass.

**Comment by lwDavid** [comment]
Reviewed `858c72e` against its merge-base `e1e15bc`, and re-ran the affected paths on a 238 A5 box (CI-pinned toolchain, ptoas 0.61).

## The core change holds up

I hand-simulated the kernel, `compressor_ratio2_paged`, `official_reference_c2a` and `_kernel_golden` across chunk lengths below / at / above capacity, odd- and even-start chunks, capacity 1/2/4/7, empty requests, reversed block maps and out-of-range blocks. They agree bit-for-bit under the documented contract. The load-bearing claim — that writing only the last `min(chunk, STATE_CAPACITY)` rows leaves the same final ring as writing every token — is true for any chunk length, because consecutive positions make those rows the last writer of each distinct residue.

Mutation testing, to check the tests actually bite:

| mutation | result |
| --- | --- |
| `block = request` (i.e. revert to batch-row addressing) | a5sim **FAIL** |
| drop `t >= end - STATE_CAPACITY` in the kernel | a5sim **FAIL** |
| `t > start` → `t >= start` in `compressor_pair` | a5sim **FAIL** |
| `(position - 1) % cap` → `(position + 1) % cap` | pytest **8 failed** |
| drop `t >= end - capacity` in `compressor_ratio2_paged` | 17 passed — correct, it is an equivalent mutant |

The decode fixture is itself a regression test for the batch-order bug: `state_block_table = arange(requests-1,-1,-1)+1` is a deliberately reversed map and prefixes `(1, 63, 128, 255, 0, 511)` start three requests at an odd position with a seeded ring row, so a reversion fails the gating A5 job immediately. The prefill fixture's chunks `[11, 8, 8, 8, 8, 5]` all exceed capacity, so the multi-writer window is exercised on device under the new untouched-byte check.

Also ran: 17 CPU contract tests, 405 repo unit tests, the a5sim two-call case including the empty-work path, `--case masked` (every state block `-1`) on device, and `--epochs 4` three times on both arms — all pass.

No measurable cost. `decode_c2a_full -p a5 --tp 1`, `effective_us` min over three reps each:

```
base e1e15bc   3888.0  3912.1  3938.1
this PR        3966.6  3952.0  3879.1
```

Best-of-3 min 3888.0 vs 3879.1 — the run-to-run spread swamps any difference, so the extra scalar reads in `compressor_pool` and `compressor_pair` do not show up end to end.

For the record, the `pl.write(parity, [lane, 0], 1.0)` flag in the automated review is a false positive: `pl.full` yields a non-GM tensor, `tensor.write` on a non-GM destination is rewritten to `tile.write` and lowers to `tsetval` with a flat offset, and `prefill_c1a_common` already does the same thing on main.

## Worth fixing

**1. The documented sim command does not run.** The new docs section says:

```
python tests/contract/test_v41_compressor_state.py -p a5sim
```

which fails with `ModuleNotFoundError: No module named 'models'`. Every runnable model entry carries a `sys.path.insert(0, str(Path(__file__).resolve().parents[2]))` bootstrap; this file has none and relies on `tests/contract/conftest.py`, which only pytest loads. Three lines fix it, and I would rather have that than an invitation that fails on first use.

**2. Cross-dispatch state has no automated guard.** Single-dispatch addressing is gated on device as above, but departure / re-admission / released-block reuse — two separate launches sharing one ring — lives only in the `__main__` half of the new test, and nothing runs it: `unit-tests` runs `pytest tests` (CPU only), and `a5-v41-flash` selects entries by the `# ci: a5` marker under `models/`. That is the part of the contract with no coverage at all.

## Worth a line of doc

**3. `STATE_CAPACITY = 4` is below `DECODE_ROWS_PER_REQUEST = 6`.** Measured against the unchunked reference: a 6-row speculative step with 2 drafts accepted resumes on a predecessor the same call already clobbered (position 2 aliases position 6 at capacity 4) and pools the wrong pair; capacity 5, 6 and 8 are all correct. This is not a regression — the old layout was effectively capacity 1 — and no capacity helps a 4096-token prefill chunk, so raising the constant is optional. But the doc currently says rollback is the engine's responsibility without noting that the ring cannot supply the bytes, so a future DSpark engine must resume from the preceding even position rather than a rejected draft boundary. One sentence, or bump the constant to 6.

## Minor

**4.** `compile_runtime=True` on `num_tokens` is not mentioned in the PR body, and it de-specializes every spmd grid and tile loop in both C2A entries, not just the compressor. The `if num_tokens > 0` guard it exists to enable cannot be reached through either entry — both `parser.error` unless `--tokens >= 1` — so the zero-work path ships untested. Fine by measurement, but it deserves a line in the body.

**5.** `compressor_ratio2` returns `state_tid` and both production call sites drop it, so the ring write is ordered against the next epoch's ring read only by the `c2a_previous_epoch` output fence. I could not reproduce a failure (`--epochs 4`, three runs per arm, bit-stable `rel_l2`), but the dependency is expressed nowhere — either thread it through or drop the return value.

**6.** The `pl.MX_B_NN` per-rank annotations are a real gap-fill: `prefill_c2a_full` and `prefill_c2a_reuse` are the only two entries on main still missing them after #1297. Correct and wanted, but unrelated to compressor state — worth its own PR, or at least its own line in the body.
